# Capstone 1 Step 2 Data Wrangling


Corona Virus disease (COVID-19) is an infectious disease caused by a newly discovered virus, which emerged in Wuhan, China in December of 2019.


<img src="figures/covid.png" />
Most people infected with the COVID-19 virus will experience mild to moderate respiratory illness and recover without requiring special treatment.  Older people and those with underlying medical problems like cardiovascular disease, diabetes, chronic respiratory disease, and cancer are more likely to develop serious illness.


The COVID-19 virus spreads primarily through droplets of saliva or discharge from the nose when an infected person coughs or sneezes, so you might have heard caution to practice respiratory etiquette (for example, by coughing into a flexed elbow).
This project is about the prediction of covid-19 cases in upcoming days and the factors contributing to the drastic spread of virus.



<font color = 'teal'>Data wrangling is the second step in Data Science methods. Data Wrangling includes Data Collection, Data Organisation, Data Definition and Data Cleaning.</font>

# Data Collection

<font color='teal'>Data Collection includes importing the packages and modules in python. And it includes loading the data in pandas dataframe.</font>



In [1]:
#importing libraries
import pandas as pd
import numpy as np
import datetime
import matplotlib.pyplot as plt
import seaborn as sns 
import os
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
warnings.simplefilter('ignore')

In [2]:
#Printing the current working directory
os.getcwd()

'C:\\Users\\user\\Springboard\\Course\\CapStone - 1\\Data Wrangling'

In [3]:
#printing the files in the current working directory
os.listdir()

['.ipynb_checkpoints',
 'Capstone Project - Data Wrangling .ipynb',
 'covid_data.csv',
 'covid_var_desc.csv',
 'data',
 'figures',
 'models']

<font color='teal'>Loading the data from csv file into a pandas dataframe df.</font>

In [4]:
df = pd.read_csv('covid_data.csv')
#print the first five rows of the dataframe
df.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2019-12-31          0.0        0.0   
1      AFG      Asia  Afghanistan  2020-01-01          0.0        0.0   
2      AFG      Asia  Afghanistan  2020-01-02          0.0        0.0   
3      AFG      Asia  Afghanistan  2020-01-03          0.0        0.0   
4      AFG      Asia  Afghanistan  2020-01-04          0.0        0.0   

   total_deaths  new_deaths  total_cases_per_million  new_cases_per_million  \
0           0.0         0.0                      0.0                    0.0   
1           0.0         0.0                      0.0                    0.0   
2           0.0         0.0                      0.0                    0.0   
3           0.0         0.0                      0.0                    0.0   
4           0.0         0.0                      0.0                    0.0   

   ...  aged_70_older  gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
0  ...          1.337        1803.987              NaN                597.029   
1  ...          1.337        1803.987              NaN                597.029   
2  ...          1.337        1803.987              NaN                597.029   
3  ...          1.337        1803.987              NaN                597.029   
4  ...          1.337        1803.987              NaN                597.029   

   diabetes_prevalence  female_smokers  male_smokers  handwashing_facilities  \
0                 9.59             NaN           NaN                  37.746   
1                 9.59             NaN           NaN                  37.746   
2                 9.59             NaN           NaN                  37.746   
3                 9.59             NaN           NaN                  37.746   
4                 9.59             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  
0                         0.5            64.83  
1                         0.5            64.83  
2                         0.5            64.83  
3                         0.5            64.83  
4                         0.5            64.83  

[5 rows x 36 columns]

# Data Organisation

<font color = 'teal'> Data Organisation : This step of Data Science invloves creating sub folders to ensure the project is in well organised manner.</font>

<font color = 'teal'>Creating sub-folders</font>

In [ ]:
#create sub-folders  named data
os.mkdir('data')

In [ ]:
#create sub-folders  named data
os.mkdir('models')

In [ ]:
#Create sub-folder named figures for saving the visualisation in the future steps
os.mkdir('figures')

In [5]:
#Making sure the sub-folders are created
os.listdir()

['.ipynb_checkpoints',
 'Capstone Project - Data Wrangling .ipynb',
 'covid_data.csv',
 'covid_var_desc.csv',
 'data',
 'figures',
 'models']

# Data Definition
Understanding the dataset

<font color = 'teal'>Data Definition includes defining the data such as coloumn names, dtypes of the coloumn, description of the coloumn, count of unique values or codes and range of unique values or codes including NAN values</font>

In [6]:
#priting the coloumns of the dataframe df
df.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_cases_per_million',
       'new_cases_per_million', 'total_deaths_per_million',
       'new_deaths_per_million', 'new_tests', 'total_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'tests_per_case', 'positive_rate', 'tests_units', 'stringency_index',
       'population', 'population_density', 'median_age', 'aged_65_older',
       'aged_70_older', 'gdp_per_capita', 'extreme_poverty',
       'cardiovasc_death_rate', 'diabetes_prevalence', 'female_smokers',
       'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand',
       'life_expectancy'],
      dtype='object')

<font color = 'teal'>Describing each column of the dataframe df</font>

In [7]:
col_desc = pd.read_csv("covid_var_desc.csv")
col_desc.set_index('column')
col_desc

column  \
0                          iso_code   
1                         continent   
2                          location   
3                              date   
4                       total_cases   
5                         new_cases   
6                      total_deaths   
7                        new_deaths   
8           total_cases_per_million   
9             new_cases_per_million   
10         total_deaths_per_million   
11           new_deaths_per_million   
12                      total_tests   
13                        new_tests   
14               new_tests_smoothed   
15         total_tests_per_thousand   
16           new_tests_per_thousand   
17  new_tests_smoothed_per_thousand   
18                   tests_per_case   
19                    positive_rate   
20                      tests_units   
21                 stringency_index   
22                       population   
23               population_density   
24                       median_age   
25                    aged_65_older   
26                    aged_70_older   
27                   gdp_per_capita   
28                  extreme_poverty   
29            cardiovasc_death_rate   
30              diabetes_prevalence   
31                   female_smokers   
32                     male_smokers   
33           handwashing_facilities   
34       hospital_beds_per_thousand   
35                  life_expectancy   

                                          description  \
0     ISO 3166-1 alpha-3 – three-letter country codes   
1              Continent of the geographical location   
2                               Geographical location   
3                                 Date of observation   
4                   Total confirmed cases of COVID-19   
5                     New confirmed cases of COVID-19   
6                 Total deaths attributed to COVID-19   
7                   New deaths attributed to COVID-19   
8   Total confirmed cases of COVID-19 per 1,000,00...   
9   New confirmed cases of COVID-19 per 1,000,000 ...   
10  Total deaths attributed to COVID-19 per 1,000,...   
11  New deaths attributed to COVID-19 per 1,000,00...   
12                           Total tests for COVID-19   
13                             New tests for COVID-19   
14  New tests for COVID-19 (7-day smoothed). For c...   
15          Total tests for COVID-19 per 1,000 people   
16            New tests for COVID-19 per 1,000 people   
17  New tests for COVID-19 (7-day smoothed) per 1,...   
18  Tests conducted per new confirmed case of COVI...   
19  The share of COVID-19 tests that are positive,...   
20  Units used by the location to report its testi...   
21  Government Response Stringency Index: composit...   
22                                 Population in 2020   
23  Number of people divided by land area, measure...   
24  Median age of the population, UN projection fo...   
25  Share of the population that is 65 years and o...   
26  Share of the population that is 70 years and o...   
27  Gross domestic product at purchasing power par...   
28  Share of the population living in extreme pove...   
29  Death rate from cardiovascular disease in 2017...   
30  Diabetes prevalence (% of population aged 20 t...   
31  Share of women who smoke, most recent year ava...   
32  Share of men who smoke, most recent year avail...   
33  Share of the population with basic handwashing...   
34  Hospital beds per 1,000 people, most recent ye...   
35                   Life expectancy at birth in 2019   

                                               source  
0      International Organization for Standardization  
1                                   Our World in Data  
2                                   Our World in Data  
3                                   Our World in Data  
4   European Centre for Disease Prevention and Con...  
5   European Centre for Disease Prevention and Con...  
6   European Centre for Disease Prevention and Con...  
7   European Centre for Dis

In [8]:
#printing the dimensions of the dataframe df
df.shape

(35717, 36)

In [9]:
#printing the dtypes of the dataframe df
df.dtypes

iso_code                            object
continent                           object
location                            object
date                                object
total_cases                        float64
new_cases                          float64
total_deaths                       float64
new_deaths                         float64
total_cases_per_million            float64
new_cases_per_million              float64
total_deaths_per_million           float64
new_deaths_per_million             float64
new_tests                          float64
total_tests                        float64
total_tests_per_thousand           float64
new_tests_per_thousand             float64
new_tests_smoothed                 float64
new_tests_smoothed_per_thousand    float64
tests_per_case                     float64
positive_rate                      float64
tests_units                         object
stringency_index                   float64
population                         float64
population_

In [10]:
#using info() to review the coloumn names, no of null values in all the columns and their dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35717 entries, 0 to 35716
Data columns (total 36 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   iso_code                         35653 non-null  object 
 1   continent                        35430 non-null  object 
 2   location                         35717 non-null  object 
 3   date                             35717 non-null  object 
 4   total_cases                      35372 non-null  float64
 5   new_cases                        35372 non-null  float64
 6   total_deaths                     35372 non-null  float64
 7   new_deaths                       35372 non-null  float64
 8   total_cases_per_million          35308 non-null  float64
 9   new_cases_per_million            35308 non-null  float64
 10  total_deaths_per_million         35308 non-null  float64
 11  new_deaths_per_million           35308 non-null  float64
 12  new_tests         

<font color = 'teal'>Print the count of unique values in each column</font>

In [11]:
df.nunique()

iso_code                             211
continent                              6
location                             212
date                                 223
total_cases                        12039
new_cases                           3193
total_deaths                        4139
new_deaths                           887
total_cases_per_million            21792
new_cases_per_million              12063
total_deaths_per_million           10660
new_deaths_per_million              2606
new_tests                           7122
total_tests                        11028
total_tests_per_thousand            9141
new_tests_per_thousand              2284
new_tests_smoothed                  7917
new_tests_smoothed_per_thousand     2305
tests_per_case                     10636
positive_rate                        501
tests_units                            5
stringency_index                     160
population                           211
population_density                   200
median_age      

<font color = 'teal'>Printing the range of values in each column</font>

In [12]:
#iterate over the columns of the dataframe by dropping the columns whose dtypes are object
#print the range for all the columns
for col in df.drop(columns=['iso_code','continent','location','tests_units']).columns:
    print('column_name: ',col,'min: ',min(df[col]),'max: ',max(df[col]))

column_name:  date min:  2019-12-31 max:  2020-08-09
column_name:  total_cases min:  0.0 max:  19624044.0
column_name:  new_cases min:  -2461.0 max:  298083.0
column_name:  total_deaths min:  0.0 max:  726953.0
column_name:  new_deaths min:  -1918.0 max:  10504.0
column_name:  total_cases_per_million min:  0.0 max:  39100.192
column_name:  new_cases_per_million min:  -265.189 max:  4944.376
column_name:  total_deaths_per_million min:  0.0 max:  1237.5510000000002
column_name:  new_deaths_per_million min:  -41.023 max:  200.04
column_name:  new_tests min:  nan max:  nan
column_name:  total_tests min:  nan max:  nan
column_name:  total_tests_per_thousand min:  nan max:  nan
column_name:  new_tests_per_thousand min:  nan max:  nan
column_name:  new_tests_smoothed min:  nan max:  nan
column_name:  new_tests_smoothed_per_thousand min:  nan max:  nan
column_name:  tests_per_case min:  nan max:  nan
column_name:  positive_rate min:  nan max:  nan
column_name:  stringency_index min:  nan max: 

In [13]:
#to know how many countries are there in the dataframe df
df['location'].value_counts()

France           223
Singapore        223
China            223
South Korea      223
Taiwan           223
                ... 
Tajikistan       101
Comoros          100
Lesotho           87
International     64
Hong Kong         11
Name: location, Length: 212, dtype: int64

In [14]:
#use describe() to get the summary statistics
df.describe()

total_cases      new_cases   total_deaths    new_deaths  \
count  3.537200e+04   35372.000000   35372.000000  35372.000000   
mean   5.893766e+04    1109.580685    2915.338007     41.103302   
std    6.042361e+05   10290.585675   27532.892118    354.836792   
min    0.000000e+00   -2461.000000       0.000000  -1918.000000   
25%    2.400000e+01       0.000000       0.000000      0.000000   
50%    5.420000e+02       6.000000      10.000000      0.000000   
75%    6.230750e+03     115.000000     121.000000      2.000000   
max    1.962404e+07  298083.000000  726953.000000  10504.000000   

       total_cases_per_million  new_cases_per_million  \
count             35308.000000           35308.000000   
mean               1227.542967              18.981592   
std                2891.630287              64.023822   
min                   0.000000            -265.189000   
25%                  10.831250               0.000000   
50%                 190.196000               0.909500   
75%                1084.090000              11.760000   
max               39100.192000            4944.376000   

       total_deaths_per_million  new_deaths_per_million      new_tests  \
count              35308.000000            35308.000000   11089.000000   
mean                  43.964543                0.540077   17750.008387   
std                  127.040536                2.950397   65353.264565   
min                    0.000000              -41.023000   -3743.000000   
25%                    0.000000                0.000000     848.000000   
50%                    2.557000                0.000000    2939.000000   
75%                   25.822750                0.154000   10214.000000   
max                 1237.551000              200.040000  929838.000000   

        total_tests  ...  aged_70_older  gdp_per_capita  extreme_poverty  \
count  1.143600e+04  ...   31944.000000    31711.000000     21214.000000   
mean   8.955480e+05  ...       5.960218    21399.674824        11.633398   
std    3.598348e+06  ...       4.352825    20652.863952        18.851105   
min    1.000000e+00  ...       0.526000      661.240000         0.100000   
25%    3.064575e+04  ...       2.162000     5819.495000         0.500000   
50%    1.281800e+05  ...       4.455000    14600.861000         1.800000   
75%    5.029558e+05  ...       9.720000    32605.906000        15.000000   
max    5.965268e+07  ...      18.493000   116935.600000        77.600000   

       cardiovasc_death_rate  diabetes_prevalence  female_smokers  \
count           32135.000000         33217.000000    25434.000000   
mean              249.907645             8.040270       10.957627   
std               117.841040             4.125334       10.499695   
min                79.370000             0.990000        0.100000   
25%               153.493000             5.310000        1.900000   
50%               235.954000             7.110000        6.434000   
75%               318.949000            10.180000       19.600000   
max               724.417000            23.360000       44.000000   

       male_smokers  handwashing_facilities  hospital_beds_per_thousand  \
count  25126.000000            14776.000000                29174.000000   
mean      32.628051               53.047610                    3.138716   
std       13.344887               31.496675                    2.543218   
min        7.700000                1.188000                    0.100000   
25%       21.400000               22.863000                    1.380000   
50%       31.400000               55.182000                    2.540000   
75%       40.900000               83.741000                    4.210000   
max       78.100000               98.999000                   13.800000   

       life_expectancy  
count     35218.000000  
mean         74.190996  
std           7.329327  
min          53.280000  
25%          70.390000  
50%          75.800000  
75%          80.070000  
max          86.750000  

[8 rows x 31 c

The dataframe df consists of 35717 rows and 36 features having information about the covid-19 cases in different countries from december,2019 till August 9th,2020. 

# Data Cleaning

The data originally obtained was in csv file and directly loaded into the pandas dataframe effortlessly. In other words, the dataset we have in our hands is already relatively clean. We will however attempt at learning more about our features and performing appropriate cleaning steps to arrive at a form that is more suitable for analysis.

<font color = 'teal'>Data cleaning is the process of handling missing values(NAN) and look into duplicate rows</font>

In [15]:
df[['iso_code','location']]

iso_code       location
0          AFG    Afghanistan
1          AFG    Afghanistan
2          AFG    Afghanistan
3          AFG    Afghanistan
4          AFG    Afghanistan
...        ...            ...
35712      NaN  International
35713      NaN  International
35714      NaN  International
35715      NaN  International
35716      NaN  International

[35717 rows x 2 columns]

 The ISO country codes are internationally recognized codes that designate every country and most of the dependent areas a two-letter combination or a three-letter combination; it is like an acronym, that stands for a country or a state. Since both column represent the same country or state, dropping iso_code from the dataframe df would do.

In [16]:
#Drop the iso_code column from the dataframe df using df.drop()
df=df.drop(columns=['iso_code'],axis=1)
#Printing the first five rows
df.head()

continent     location        date  total_cases  new_cases  total_deaths  \
0      Asia  Afghanistan  2019-12-31          0.0        0.0           0.0   
1      Asia  Afghanistan  2020-01-01          0.0        0.0           0.0   
2      Asia  Afghanistan  2020-01-02          0.0        0.0           0.0   
3      Asia  Afghanistan  2020-01-03          0.0        0.0           0.0   
4      Asia  Afghanistan  2020-01-04          0.0        0.0           0.0   

   new_deaths  total_cases_per_million  new_cases_per_million  \
0         0.0                      0.0                    0.0   
1         0.0                      0.0                    0.0   
2         0.0                      0.0                    0.0   
3         0.0                      0.0                    0.0   
4         0.0                      0.0                    0.0   

   total_deaths_per_million  ...  aged_70_older  gdp_per_capita  \
0                       0.0  ...          1.337        1803.987   
1                       0.0  ...          1.337        1803.987   
2                       0.0  ...          1.337        1803.987   
3                       0.0  ...          1.337        1803.987   
4                       0.0  ...          1.337        1803.987   

   extreme_poverty  cardiovasc_death_rate  diabetes_prevalence  \
0              NaN                597.029                 9.59   
1              NaN                597.029                 9.59   
2              NaN                597.029                 9.59   
3              NaN                597.029                 9.59   
4              NaN                597.029                 9.59   

   female_smokers  male_smokers  handwashing_facilities  \
0             NaN           NaN                  37.746   
1             NaN           NaN                  37.746   
2             NaN           NaN                  37.746   
3             NaN           NaN                  37.746   
4             NaN           NaN                  37.746   

   hospital_beds_per_thousand life_expectancy  
0                         0.5           64.83  
1                         0.5           64.83  
2                         0.5           64.83  
3                         0.5           64.83  
4                         0.5           64.83  

[5 rows x 35 columns]

<font color = 'teal'>Handling missing values</font>

In [17]:
nan = pd.DataFrame(df.isnull().sum().sort_values(ascending=False)/len(df),columns = ['Percent'])
nan_greaterthan_0 = nan['Percent'] > 0 
nan[nan_greaterthan_0]

Percent
new_tests_per_thousand           0.689532
new_tests                        0.689532
total_tests_per_thousand         0.679816
total_tests                      0.679816
tests_per_case                   0.674497
positive_rate                    0.664865
new_tests_smoothed_per_thousand  0.650923
new_tests_smoothed               0.650923
tests_units                      0.630036
handwashing_facilities           0.586303
extreme_poverty                  0.406053
male_smokers                     0.296525
female_smokers                   0.287902
stringency_index                 0.188734
hospital_beds_per_thousand       0.183190
aged_65_older                    0.114119
gdp_per_capita                   0.112159
aged_70_older                    0.105636
median_age                       0.100988
cardiovasc_death_rate            0.100288
diabetes_prevalence              0.069995
population_density               0.045581
life_expectancy                  0.013971
new_deaths_per_million           0.011451
total_deaths_per_million         0.011451
new_cases_per_million            0.011451
total_cases_per_million          0.011451
new_deaths                       0.009659
total_deaths                     0.009659
new_cases                        0.009659
total_cases                      0.009659
continent                        0.008035
population                       0.001792

In [18]:
df.drop(df[(df['location']=='International') | (df['location']=='World')].index,inplace = True)

In [19]:
# using fillna() forward fill method replacing the nan values in the column stringency_index.
#Since the stricteness of lockdown will remain same as the before day mostly
#And the first day strictness level maybe 0. So filling the first row with 0 since applying ffill will not work for the first row.
df["stringency_index"].fillna( method ='ffill', inplace = True) 
df[:1].fillna(0, inplace=True)

In [20]:
#using fillna(), filling the nan values of 'extreme_poverty' by the  column mean.
#And using groupby() grouping the dataframe with location column
df['extreme_poverty'] = df.groupby(['location'])['extreme_poverty'].fillna(df.extreme_poverty.mean())

In [21]:
#using fillna(), filling the nan values of 'population_density' by the column mean
#Aand using groupby() to group the dataframe by location column
df['population_density'] = df.groupby(['location'])['population_density'].fillna(df.population.mean())

In [22]:
#using fillna(), filling the nan values of 'median_age' by the  column mean and using groupby location
df['median_age'] = df.groupby(['location'])['median_age'].fillna(df.median_age.mean())

In [23]:
#using fillna(), filling the nan values of 'aged_65_older' by the  column mean and using groupby location
df['aged_65_older'] = df.groupby(['location'])['aged_65_older'].fillna(df.aged_65_older.mean())

In [24]:
#using fillna(), filling the nan values of 'aged_70_older' by the  column mean and using groupby location
df['aged_70_older'] = df.groupby(['location'])['aged_70_older'].fillna(df.aged_65_older.mean())

In [25]:
#using fillna(), filling the nan values of 'gdp_per_capita' by the  column mean and using groupby location
df['gdp_per_capita'] = df.groupby(['location'])['gdp_per_capita'].fillna(df.gdp_per_capita.mean())

In [26]:
#using fillna(), filling the nan values of 'diabetes_prevalence' by the  column mean and using groupby location
df['diabetes_prevalence'] = df.groupby(['location'])['diabetes_prevalence'].fillna(df.diabetes_prevalence.mean())

In [27]:
#using fillna(), filling the nan values of 'cvd_death_rate' by the  column mean and using groupby location
df['cardiovasc_death_rate'] = df.groupby(['location'])['cardiovasc_death_rate'].fillna(df.cardiovasc_death_rate.mean())

In [28]:
#using fillna(), filling the nan values of 'female_smokers' by the  column mean and using groupby location
df['female_smokers'] = df.groupby(['location'])['female_smokers'].fillna(0)

In [29]:
#using fillna(), filling the nan values of 'male_smokers' by the  column mean and using groupby location
df['male_smokers'] = df.groupby(['location'])['male_smokers'].fillna(0)

In [30]:
#using fillna(), filling the nan values of 'life_expectancy' by the  column mean and using groupby location
df['life_expectancy'] = df.groupby(['location'])['life_expectancy'].fillna(df.life_expectancy.mean())

In [31]:
df['hospital_beds_per_thousand'] = df.groupby(['location'])['hospital_beds_per_thousand'].fillna(0)

In [32]:
df['total_cases'] = df.groupby(['location'])['total_cases'].fillna(method = 'bfill')
df['total_cases'] = df.groupby(['location'])['total_cases'].fillna(0)

In [33]:
df['new_cases'] = df.groupby(['location'])['new_cases'].fillna(0)

In [34]:
df['total_deaths'] = df.groupby(['location'])['total_deaths'].fillna(method = 'bfill')
df['total_deaths'] = df.groupby(['location'])['total_deaths'].fillna(0)

In [35]:
df['new_deaths'] = df.groupby(['location'])['new_deaths'].fillna(0)

In [36]:
df['total_cases_per_million'] = df.groupby(['location'])['total_cases_per_million'].fillna(method = 'bfill')
df['total_cases_per_million'] = df.groupby(['location'])['total_cases_per_million'].fillna(0)

In [37]:
df['new_cases_per_million'] = df.groupby(['location'])['new_cases_per_million'].fillna(0)

In [38]:
df['total_deaths_per_million'] = df.groupby(['location'])['total_deaths_per_million'].fillna(method = 'bfill')
df['total_deaths_per_million'] = df.groupby(['location'])['total_deaths_per_million'].fillna(0)

In [39]:
df['handwashing_facilities'] = df.groupby(['location'])['handwashing_facilities'].fillna(method = 'bfill')
df['handwashing_facilities'] = df.groupby(['location'])['handwashing_facilities'].fillna(0)

In [40]:
df['new_deaths_per_million'] = df.groupby(['location'])['new_deaths_per_million'].fillna(0)

In [41]:
df['tests_per_case']=df['tests_per_case'].interpolate()

In [42]:
df['positive_rate']=df['positive_rate'].interpolate()

In [43]:
df['new_tests']=df['new_tests'].interpolate()

In [44]:
df['total_tests']=df['total_tests'].interpolate()

In [45]:
df['total_tests_per_thousand']=df['total_tests_per_thousand'].interpolate()

In [46]:
df['new_tests_per_thousand']=df['new_tests_per_thousand'].interpolate()

In [47]:
df['new_tests_smoothed']=df['new_tests_smoothed'].interpolate()

In [48]:
df['new_tests_smoothed_per_thousand']=df['new_tests_smoothed_per_thousand'].interpolate()

In [49]:
df['tests_units']=df['tests_units'].fillna(0)

<font color = 'teal'> Look for duplicate rows </font>

In [50]:
df_duplicated = df[df.duplicated()]
df_duplicated

Empty DataFrame
Columns: [continent, location, date, total_cases, new_cases, total_deaths, new_deaths, total_cases_per_million, new_cases_per_million, total_deaths_per_million, new_deaths_per_million, new_tests, total_tests, total_tests_per_thousand, new_tests_per_thousand, new_tests_smoothed, new_tests_smoothed_per_thousand, tests_per_case, positive_rate, tests_units, stringency_index, population, population_density, median_age, aged_65_older, aged_70_older, gdp_per_capita, extreme_poverty, cardiovasc_death_rate, diabetes_prevalence, female_smokers, male_smokers, handwashing_facilities, hospital_beds_per_thousand, life_expectancy]
Index: []

[0 rows x 35 columns]

From the above output we can ensure that there are no duplicate rows in the dataframe df

# Export the dataframe to a new csv file 

<font color = 'teal'>Export the processed dataframe as a csv file into the data folder which we created in data organisation step.</font>

In [51]:
df.to_csv('C:/Users/user/Springboard/Course/CapStone - 1/Exploratory Data Analysis/covid_updated_file.csv',index=True)